enhancements/operator-dependency-resolution.md

```diff
@@ -0,0 +1,222 @@
+
+* The data pulled for the resolution is distributed across several catalogs in several kube namespaces. As a result, we need to weight the dependencies so that we can maintain the preference of same catalog, then same namespace for dependencies
+
+* Replacement chains and update graphs (channels) are constraints. For a fresh install, we can prefer the default channel wherever possible. If an existing subscription owns a dependency, we will constrain the resolution to only consider versions in the channel for the existing subscription. We will also need to build the resolution incrementally in the case that the catalogs update and several new updates are available, stepping through each option and solving for each incremental update.
+
+* Existing operator group and install modes. Once the operator group is created, we should only consider installing operators that have compatible install modes.
+
+* Error messaging for when the resolution is unsatisfiable (due to missing dependencies or dependency conflicts) is something we need to address. See -- https://en.wikipedia.org/wiki/Resolution_(logic)#Propositional_non-clausal_resolution_example. Additionally, we will need to ensure that error messaging is bubbled up to the subscription and operator API.
+
+#### SAT Solver
+
+The implementation of a dependency resolver is a satisfiability problem and NP-Complete. As such, the runtime of a trivial resolver becomes 2^n. For an initial implementation to have relatively good average runtime, we need to transform the dependency problem into a set of boolean statements and feed them into a SAT solver algorithm. Rather than implement one from scratch, our proposal is to use a well tested existing SAT solver library to solve this problem. However, because dependency resolution is a relatively narrow field, there does not exist a generic dependency resolver that would be trivial for our project to import. Instead, we have evaluated several existing solver libraries and compared them with our own constraints and intend to wrap that solver in our own resolver library to satisfy all of them.
+
+Imported SAT solver selection constraints:
+* Written in pure Go
+* Large project with a healthy ecosystem, good documentation and recent commits
+* Common interface that would be easy to replace if needed
+* Open Source license
+
+SAT solver needs to take in account several other constraints in order to resolve dependencies correctly. Here is the list of potential resolution constraints:
+* InstallMode (OperatorGroup)
+* Minimum Kubernetes version
+* Native APIs
+* OpenShift version (OpenShift only)
+
+After evaluating several libraries, we have chosen https://github.com/irifrance/gini as the SAT solver. This implementation of the solver is general, a popular library, and is written in pure go.
```

Comment:
The main driver is that it's the only pure go solver library that seems remotely active and is being used by other projects. Our other options are basically:

https://github.com/mitchellh/go-sat -- last commit 3 years ago
https://github.com/crillab/gophersat -- this isn't really a library, it's an academic project used to solve SAT problems in DIMACS format. it technically has some lib functions, but they aren't documented and it's not the projects intended use. Seems like there's more recent active development though (this year)

This has interoperability with c: https://github.com/justinfx/pigosat
but it's basically just someone's pet project. Had a nice wrapper for dealing with version dependencies though: https://github.com/justinfx/pakr

And the only other solver I found that seemed somewhat reasonable is the one `dep` uses: https://github.com/golang/dep/tree/master/gps

But that thing is really tightly coupled with the way that dep does resolution so we would have to basically fork it and maintain it ourselves.

Reply:
> But that thing is really tightly coupled with the way that dep does resolution so we would have to basically fork it and maintain it ourselves.

Writing or forking and maintaining our own solver may be a good way to get the best performance or properties (like really great error messages) long term. We should wrap whatever we do choose with a good interface that we can potentially swap in the future.

Reply:
Yep, agreed. I think another good reason to use the library we chose is because its interface is really simple (basically, just list all the boolean constraints and then `solver.Solve()`). If we need to write our own solver the easiest part will certainly be replicating that interface.